sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala

```diff
@@ -303,6 +295,66 @@ case class Hex(child: Expression) extends UnaryExpression with Serializable  {
   }
 }
 
+/**
+ * Performs the inverse operation of HEX.
+ * Resulting characters are returned as a byte array.
+ */
+case class Unhex(child: Expression)
+  extends UnaryExpression with AutoCastInputTypes with Serializable {
+
+  override def nullable: Boolean = true
+  override def dataType: DataType = BinaryType
+  override def inputTypes: Seq[DataType] = Seq(BinaryType)
+
+  override def eval(input: InternalRow): Any = {
+    val num = child.eval(input)
+    if (num == null) {
+      null
+    } else {
+      unhex(num.asInstanceOf[UTF8String].getBytes)
+    }
+  }
+
+  // lookup table to translate '0' -> 0 ... 'F'/'f' -> 15
+  private[this] val unhexDigits = {
```

Comment:
BTW I absolutely disagree that "less code is better than more code" as an ethos. While it can be true in many cases, it is often also not true. Plenty of counter examples:

- putting everything into a single file without any namespacing reduces the import statements, but that creates bad logical structures
- not writing any test cases reduces the amount of code also, but that's obviously bad
- using arcane scala features can substantially reduce code size in certain cases, at the cost of readability

In this case, I don't see how this creates less code (it takes 2 lines of code to define a scala object -- you can even put it in an existing java class like in UTF8String as a static field).